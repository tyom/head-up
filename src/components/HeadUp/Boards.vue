<template>
  <div class="HeadUpBoards flex-grow overflow-hidden">
    <Board
      v-for="(board, idx) in boards"
      v-bind="board"
      :key="board.title + board.id + idx"
    />
    <slot />
  </div>
</template>

<script>
import Board from '../Board';

export default {
  components: {
    Board,
  },
  props: {
    boards: {
      type: Array,
      default: () => [],
    },
  },
};
</script>
