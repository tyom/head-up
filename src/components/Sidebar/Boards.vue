<template>
  <div class="Boards">
    <transition-group tag="ul" name="boardReveal" class="board-list">
      <li
        v-for="board in boards"
        :key="board.id"
        :class="getBoardListItemClass(board)"
        :title="getBoardTitle(board)"
        class="relative"
      >
        <transition name="actionsReveal">
          <div v-if="editMode && board.editable" class="board-actions">
            <button
              type="button"
              class="remove-button"
              @click="$emit('remove', board.id)"
            >
              <v-icon name="minus-circle" />
            </button>
          </div>
        </transition>
        <button class="board-button" @click="$emit('activate', board.id)">
          <Board :id="board.id" :cells="board.cells" preview class="board" />
        </button>
        <div v-if="board.title" class="board-title">
          {{ board.title }}
        </div>
      </li>
    </transition-group>
  </div>
</template>

<script>
import Board from '../Board';

export default {
  components: {
    Board,
  },
  props: {
    boards: {
      type: Array,
      default: () => [],
    },
    activeId: {
      type: String,
      default: '',
    },
    editMode: {
      type: Boolean,
      default: false,
    },
  },
  methods: {
    getBoardTitle(board) {
      return board.title + (board.editable ? '' : ' (read-only)');
    },
    getBoardListItemClass(board) {
      return {
        _active: this.activeId === board.id,
        ['_read-only']: !board.editable,
      };
    },
  },
};
</script>

<style scoped>
.board-list {
  @apply grid gap-3;
}

.board-title {
  @apply text-xs text-center mt-1 -mx-2 transition duration-100;
}

.board {
  @apply transition duration-200 relative border;
  border-color: #fff4;
  min-width: 4rem;
  height: 10vh;

  &:hover {
    border-color: #fff8;
  }

  @nest ._active &,
  &:active {
    @apply border-white;
  }
}

.board-button {
  @apply block w-full p-px outline-none cursor-default;

  &.focus-visible {
    @apply shadow-outline;
  }
}

.board:hover + .board-title {
  @apply text-white;
}

._read-only .board::after {
  @apply opacity-0 absolute inset-0 transition duration-200 pointer-events-none;
  content: '';
  background: repeating-linear-gradient(
    -45deg,
    #fff0,
    #fff0 0.5rem,
    #fff1 0.5rem,
    #fff1 1rem
  );

  @nest ._edit & {
    @apply opacity-100;
  }
}

.board-actions {
  @apply absolute z-10;
  top: -0.2rem;
  right: -0.2rem;
}

.remove-button {
  @apply flex p-px bg-white rounded-full outline-none text-red-800 shadow;

  & .fa-icon {
    @apply w-3 h-3;
  }

  &:hover {
    @apply text-red-700;
  }

  &:active {
    @apply text-red-900;
  }
}

.boardReveal-enter-active,
.boardReveal-leave-active {
  transition: 0.1s;
}
.boardReveal-enter,
.boardReveal-leave-to {
  opacity: 0;
  transform: translateY(-100%);
}

.actionsReveal-enter-active,
.actionsReveal-leave-active {
  transition: 0.1s;
}
.actionsReveal-enter,
.actionsReveal-leave-to {
  transform: scale(0.1);
}
</style>
