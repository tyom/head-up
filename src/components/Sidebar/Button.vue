<template>
  <button v-bind="$attrs" type="button" class="button" v-on="$listeners">
    <slot />
  </button>
</template>

<style scoped>
button {
  @apply p-2 rounded shadow;
  background-color: #fff1;

  &:hover {
    background-color: #fff2;
  }

  &:focus {
    @apply outline-none;
  }

  &._active {
    @apply bg-gray-900 shadow-inner text-white;
  }

  &:active {
    @apply bg-blue-900;
  }

  &.focus-visible {
    @apply shadow-outline;
  }
}
</style>
