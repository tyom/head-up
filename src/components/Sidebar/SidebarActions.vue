<template>
  <div class="SidebarActions">
    <SidebarButton
      title="Shortcut: ⇧<"
      class="settings-button"
      @click="$emit('settings')"
    >
      <v-icon name="cog" />
    </SidebarButton>
    <SidebarButton
      title="Shortcut: ⇧?"
      class="help-button"
      @click="$emit('help')"
    >
      <v-icon name="question" />
    </SidebarButton>
  </div>
</template>

<script>
import SidebarButton from './SidebarButton';

export default {
  components: {
    SidebarButton,
  },
};
</script>

<style scoped>
.SidebarActions {
  display: grid;
  grid-template-columns: 1fr 1fr;
  grid-gap: 0.25em;
}
</style>
