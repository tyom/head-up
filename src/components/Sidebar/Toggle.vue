<template>
  <button class="Toggle" @click="$emit('toggle')">
    <v-icon :name="toggleIconName" />
  </button>
</template>

<script>
export default {
  props: {
    toggled: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    toggleIconName() {
      return this.toggled ? 'chevron-left' : 'ellipsis-v';
    },
  },
};
</script>

<style scoped>
.Toggle {
  @apply absolute overflow-hidden z-10 inset-0 right-auto;
  @apply w-6 h-full text-gray-500 outline-none transition duration-100 cursor-default;
  box-shadow: inset -1px 0 #fff2;

  &:hover {
    @apply text-white bg-gray-800;
  }

  &:active {
    @apply bg-blue-900;
    transform: scaleY(1.2);
  }

  &.focus-visible {
    box-shadow: inset 0 0 0 1px #fff6;
  }
}
</style>
