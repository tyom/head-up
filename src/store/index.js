import Vue from 'vue';
import Vuex from 'vuex';
import VuexPersistence from 'vuex-persist';

import actions from './actions';
import mutations from './mutations';

const vuexLocal = new VuexPersistence({
  key: 'headUp',
  saveState(key, state, storage) {
    const persistenceSetting = state.settings.persistState.value;
    if (persistenceSetting) {
      localStorage.setItem(key, JSON.stringify(state));
    } else {
      const newState = JSON.parse(storage[key]);
      newState.settings.persistState.value = persistenceSetting;
      window.localStorage.setItem(key, JSON.stringify(newState));
    }
  },
});

Vue.use(Vuex);

export const initialState = {
  activeBoardId: window.location.hash.replace(/^#/, ''),
  editMode: false,
  showSidebar: true,
  serializedBoards: [],
  settings: {
    smoothScrolling: {
      type: 'toggle',
      label: 'Smooth scrolling',
      value: true,
    },
    persistState: {
      type: 'toggle',
      label: 'Persist to local storage',
      value: true,
    },
  },
};

export default new Vuex.Store({
  actions,
  mutations,
  state: initialState,
  plugins: [vuexLocal.plugin],
});
