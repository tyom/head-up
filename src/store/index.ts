import { createStore } from 'vuex';

export const store = createStore({
  state() {
    return {
      boards: [],
    };
  },
  getters: {
    serialisedBoards(state) {
      return state.boards.map((x) => ({
        id: x.id,
        title: x.title,
      }));
    },
  },
  mutations: {
    ADD_BOARD(state, board) {
      if (state.boards.some((x) => x.id === board.id)) return;
      state.boards.push(board);
    },
  },
  actions: {
    initBoard(context, board) {
      context.commit('ADD_BOARD', board);
    },
  },
});
