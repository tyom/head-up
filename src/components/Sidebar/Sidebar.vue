<template>
  <div class="Sidebar" :class="{ _collapsed: !visible }">
    <Toggle
      :toggled="visible"
      @toggle="handleToggleSidebar"
      title="Shortcut: s"
    />
    <transition name="fade">
      <div class="wrapper" v-show="visible">
        <div class="actions">
          <Button
            class="add-button"
            title="Shortcut: a"
            @click="handleAddBoard"
          >
            <v-icon name="plus" />
          </Button>
          <Button
            :class="{ _active: editMode }"
            class="edit-button"
            title="Shortcut: e"
            @click="handleToggleEdit"
          >
            <v-icon name="edit" />
          </Button>
        </div>
        <Boards
          :edit-mode="editMode"
          :boards="boards"
          :active-id="activeBoardId"
          @activate="handleActivateBoard"
          @remove="handleRemoveBoard"
        />
        <div class="actions">
          <Button
            title="Shortcut: ⇧<"
            class="settings-button"
            @click="$emit('toggle:settings')"
          >
            <v-icon name="cog" />
          </Button>
          <Button
            title="Shortcut: ⇧?"
            class="help-button"
            @click="$emit('toggle:help')"
          >
            <v-icon name="question" />
          </Button>
        </div>
      </div>
    </transition>
  </div>
</template>

<script>
import store from '../../store';
import Toggle from './Toggle';
import Boards from './Boards';
import Button from './Button';

export default {
  components: {
    Toggle,
    Boards,
    Button,
  },
  props: {
    visible: {
      type: Boolean,
      default: false,
    },
  },
  computed: {
    boards() {
      return store.state.serializedBoards;
    },
    editMode() {
      return store.state.editMode;
    },
    activeBoardId() {
      return store.state.activeBoardId;
    },
  },
  methods: {
    handleToggleSidebar() {
      store.dispatch('TOGGLE_SIDEBAR');
    },
    handleAddBoard() {
      store.dispatch('ADD_BOARD');
    },
    handleRemoveBoard(boardId) {
      store.dispatch('REMOVE_BOARD', boardId);
    },
    handleToggleEdit() {
      store.dispatch('TOGGLE_EDIT_MODE');
    },
    handleActivateBoard(boardId) {
      store.dispatch('ACTIVATE_BOARD', boardId);
    },
  },
};
</script>

<style scoped>
.Sidebar {
  @apply relative flex overflow-hidden h-full pl-6 transition-all duration-200 ease-in;
  width: 14vw;
  min-width: 11rem;

  &._collapsed {
    @apply w-0 min-w-0;
  }
}

.wrapper {
  @apply grid flex-grow gap-4 overflow-y-auto p-4 text-gray-500;
  min-height: 100%;
  grid-template-rows: auto 1fr auto;
  background-color: #0005;
  box-shadow: inset 0 0 3px #0008;
}

.actions {
  @apply grid grid-cols-2 gap-2;
}

.fade-enter-active,
.fade-leave-active {
  @apply transition duration-700;
}

.fade-enter,
.fade-leave-to {
  @apply opacity-0;
}
</style>
