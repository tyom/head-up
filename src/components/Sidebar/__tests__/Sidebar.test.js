import { shallowMount } from '@vue/test-utils';
import store from '../../../store';
import Sidebar from '../Sidebar';
import Toggle from '../Toggle';
import Boards from '../Boards';

jest.spyOn(store, 'dispatch');

beforeEach(() => {
  jest.clearAllMocks();
});

test('render opened', () => {
  const wrapper = shallowMount(Sidebar, {
    propsData: {
      visible: true,
    },
  });
  expect(wrapper).toMatchSnapshot();
});

test('render closed', () => {
  const wrapper = shallowMount(Sidebar, {
    propsData: {
      visible: false,
    },
  });
  expect(wrapper).toMatchSnapshot();
});

describe('respond to interactions', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallowMount(Sidebar);
  });

  test('toggle sidebar', () => {
    wrapper.find(Toggle).vm.$emit('toggle');
    expect(store.dispatch).toHaveBeenCalledWith('TOGGLE_SIDEBAR');
  });

  test('add board', () => {
    wrapper.find('.add-button').vm.$emit('click');
    expect(store.dispatch).toHaveBeenCalledWith('ADD_BOARD');
  });

  test('edit mode', () => {
    wrapper.find('.edit-button').vm.$emit('click');
    expect(store.dispatch).toHaveBeenCalledWith('TOGGLE_EDIT_MODE');
  });

  test('activate board', () => {
    wrapper.find(Boards).vm.$emit('activate', 'my-board');
    expect(store.dispatch).toHaveBeenCalledWith('ACTIVATE_BOARD', 'my-board');
  });

  test('remove board', () => {
    wrapper.find(Boards).vm.$emit('remove', 'my-board');
    expect(store.dispatch).toHaveBeenCalledWith('REMOVE_BOARD', 'my-board');
  });

  test('toggle help overlay', () => {
    wrapper.find('.help-button').vm.$emit('click');
    expect(wrapper.emitted('toggle:help')[0]).toBeTruthy();
  });

  test('toggle settings overlay', () => {
    wrapper.find('.settings-button').vm.$emit('click');
    expect(wrapper.emitted('toggle:settings')[0]).toBeTruthy();
  });
});
