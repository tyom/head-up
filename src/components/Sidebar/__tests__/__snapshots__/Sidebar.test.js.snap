// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render closed 1`] = `
<div class="Sidebar _collapsed">
  <toggle-stub title="Shortcut: s"></toggle-stub>
  <transition-stub name="fade">
    <div class="wrapper" style="display: none;">
      <div class="actions">
        <button-stub title="Shortcut: a" class="add-button">
          <v-icon name="plus"></v-icon>
        </button-stub>
        <button-stub title="Shortcut: e" class="edit-button">
          <v-icon name="edit"></v-icon>
        </button-stub>
      </div>
      <boards-stub boards="" activeid=""></boards-stub>
      <div class="actions">
        <button-stub title="Shortcut: ⇧<" class="settings-button">
          <v-icon name="cog"></v-icon>
        </button-stub>
        <button-stub title="Shortcut: ⇧?" class="help-button">
          <v-icon name="question"></v-icon>
        </button-stub>
      </div>
    </div>
  </transition-stub>
</div>
`;

exports[`render opened 1`] = `
<div class="Sidebar">
  <toggle-stub toggled="true" title="Shortcut: s"></toggle-stub>
  <transition-stub name="fade">
    <div class="wrapper">
      <div class="actions">
        <button-stub title="Shortcut: a" class="add-button">
          <v-icon name="plus"></v-icon>
        </button-stub>
        <button-stub title="Shortcut: e" class="edit-button">
          <v-icon name="edit"></v-icon>
        </button-stub>
      </div>
      <boards-stub boards="" activeid=""></boards-stub>
      <div class="actions">
        <button-stub title="Shortcut: ⇧<" class="settings-button">
          <v-icon name="cog"></v-icon>
        </button-stub>
        <button-stub title="Shortcut: ⇧?" class="help-button">
          <v-icon name="question"></v-icon>
        </button-stub>
      </div>
    </div>
  </transition-stub>
</div>
`;
